proxy/agent.go

```diff
@@ -0,0 +1,198 @@
+// incremented restart epoch.  This matches Envoy semantics for restart epochs:
+// To successfully launch a new Envoy process that will replace the running
+// Envoy processes, the restart epoch of the new process must be exactly 1
+// greater than the highest restart epoch of the currently running Envoy
+// processes. The initial epoch is 0.
+//
+// Envoy hot restart documentation:
+// https://lyft.github.io/envoy/docs/intro/arch_overview/hot_restart.html
+type Agent interface {
+	// Apply sets the desired configuration for the proxy.  Agent compares the
+	// current active configuration to the desired state and initiates a restart
+	// if necessary. If the restart fails, the agent attempts to retry with an
+	// exponential back-off.
+	Apply(config interface{})
+
+	// Run starts the agent and awaits for a signal on the input channel
+	Run(stop <-chan struct{})
+}
+
+// NewAgent creates a new proxy instance agent for a "run" function is a
+// blocking call to start the proxy, "cleanup" function is called after "run"
+// finishes.  Both functions take the epoch as an argument. Retry is budgeted
+// and uses an exponential back-off from the initial delay.
+func NewAgent(run func(interface{}, int) error, cleanup func(int),
+	initialRetryBudget int, initialRetryDelay time.Duration) Agent {
+	return &agent{
+		run:                run,
+		cleanup:            cleanup,
+		epochs:             make(map[int]interface{}),
+		config:             make(chan interface{}),
+		status:             make(chan exitStatus),
+		initialRetryBudget: initialRetryBudget,
+		initialRetryDelay:  initialRetryDelay,
+	}
+}
+
+type agent struct {
+	// proxy start-up command
+	run func(interface{}, int) error
+
+	// proxy cleanup command
+	cleanup func(int)
+
+	// desired configuration state
+	desired interface{}
+
+	// active epochs and their configurations
+	epochs map[int]interface{}
+
+	// current configuration is the highest epoch configuration
+	current interface{}
+
+	// channel for posting desired configurations
+	config chan interface{}
+
+	// channel for proxy exit notifications
+	status chan exitStatus
+
+	// retry budget count
+	retryBudget int
+
+	// total budget available
+	initialRetryBudget int
+	initialRetryDelay  time.Duration
+}
+
+type exitStatus struct {
+	epoch int
+	err   error
+}
+
+func (a *agent) Apply(config interface{}) {
+	a.config <- config
+}
+
+// Run executes the reloads until stop channel closes
+func (a *agent) Run(stop <-chan struct{}) {
+	glog.V(2).Info("Starting proxy agent")
+
+	// Throttle processing up to smoothed 1 qps with bursts up to 10 qps
+	rateLimiter := flowcontrol.NewTokenBucketRateLimiter(float32(1), 10)
+
+	// Set default delay to a long duration - reconciliation is a no-op in the regular case.
+	// For permanent errors, reconciliation is still attempted once after the default delay.
+	defaultDelay := 1 * time.Hour
+	delay := defaultDelay
+
+	for {
+		rateLimiter.Accept()
```

Comment:
are we getting incremental updates or are we getting the entire config as such? If incremental updates, then this code sounds fine. But if each update is the entire config, then don't we need to just skip the earlier configs and retain just the most recent config?